jaeger-core/src/main/java/io/jaegertracing/internal/propagation/TraceContextCodec.java

```diff
@@ -0,0 +1,157 @@
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.jaegertracing.internal.propagation;
+
+import io.jaegertracing.internal.JaegerObjectFactory;
+import io.jaegertracing.internal.JaegerSpanContext;
+import io.jaegertracing.internal.utils.Utils;
+import io.jaegertracing.spi.Codec;
+import io.opentracing.propagation.TextMap;
+
+import java.util.Collections;
+import java.util.Map;
+
+
+/**
+ * This format follows the Trace Context specification  https://www.w3.org/TR/trace-context/
+ *
+ * <p>
+ * Example usage:
+ *
+ * <pre>{@code
+ * traceContextCodec = new TraceContextCodec();
+ * tracer = new JaegerTracer.Builder(serviceName, reporter, sampler)
+ *                    .registerInjector(Format.Builtin.HTTP_HEADERS, traceContextCodec)
+ *                    .registerExtractor(Format.Builtin.HTTP_HEADERS, traceContextCodec)
+ *                    ...
+ * }</pre>
+ *
+ * <p>
+ */
+public class TraceContextCodec implements Codec<TextMap> {
+  protected static final String TRACE_CONTEXT_NAME = "tracecontext";
+
+  private static final int VERSION_SIZE = 2;
+  private static final int TRACEPARENT_DELIMITER_SIZE = 1;
+  private static final int TRACE_ID_HEX_SIZE = 32;
+  private static final int SPAN_ID_HEX_SIZE = 16;
+  private static final int TRACE_ID_OFFSET = VERSION_SIZE + TRACEPARENT_DELIMITER_SIZE;
+  private static final int SPAN_ID_OFFSET =
+      TRACE_ID_OFFSET + TRACE_ID_HEX_SIZE + TRACEPARENT_DELIMITER_SIZE;
+  private static final int TRACE_OPTION_OFFSET =
+      SPAN_ID_OFFSET + SPAN_ID_HEX_SIZE + TRACEPARENT_DELIMITER_SIZE;
+  private static final int SAMPLED_FLAG = 1;
+
+  private static final byte JAEGER_SAMPLED_FLAG = 1;
+
+
+  private final JaegerObjectFactory objectFactory;
+  private final Long version;
+
+  private TraceContextCodec(Builder builder) {
+    this.objectFactory = builder.objectFactory;
+    this.version = builder.version;
+  }
+
+  @Override
+  public void inject(JaegerSpanContext spanContext, TextMap carrier) {
+    long traceIdHigh = spanContext.getTraceIdHigh();
+
+    //From the specification:
+    //When a system operates with a trace-id that is shorter than 16 bytes,
+    // it SHOULD fill-in the extra bytes with random values rather than zeroes.
+    // Let's say the system works with an 8-byte trace-id like 3ce929d0e0e4736.
+    // Instead of setting trace-id value to 0000000000000003ce929d0e0e4736,
+    // it SHOULD generate a value like 4bf92f3577b34da6a3ce929d0e0e4736.
+    // where 4bf92f3577b34da6a is a random value or a function of time and host value.
+    if (traceIdHigh == 0L) {
```

Comment:
I think now I understand. The spec does not say that you should guarantee that the traceIdHigh != 0. It is valid to have a traceIdHigh as long it was randomly generated and not set to 0.